File: Java/54_Spiral_Matrix.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;
/*
最原始的模拟法。本题无论如何都必须要访问所有元素位置，所以时间复杂度下限是O(n)
*/
public class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<Integer>();
        if(matrix==null||matrix.length==0||matrix[0].length==0){
            return res;
        }
        int row = matrix.length, col = matrix[0].length;
        int left=0,right=col,top=0,bottom=row,i,j;
        while(left<right&&top<bottom){
            for(i=left;i<right;i++){
                res.add(matrix[top][i]);
            }
            top++;
            if(top>=bottom)break;
            for(i=top;i<bottom;i++){
                res.add(matrix[i][right-1]);
            }
            right--;
            if(left>=right)break;
            for(i=right-1;i>=left;i--){
                res.add(matrix[bottom-1][i]);
            }
            bottom--;
            for(i=bottom-1;i>=top;i--){
                res.add(matrix[i][left]);
            }
            left++;
        }
        return res;
    }
}

File: python/54_Spiral_Matrix.py
# coding=utf-8
class Solution:
    # @param {integer[][]} matrix
    # @return {integer[]}
    def spiralOrder(self, matrix):
        if not matrix:
            return []
        row,col=len(matrix),len(matrix[0])
        left=top=0
        right=col-1
        down=row-1
        res=[]
        if top==down:
            return matrix[0]
        elif left==right:
            for i in range(top,down+1):
                res.append(matrix[i][right])
            return res
        while left<=right and top<=down:
            res.extend(matrix[top][left:right+1])
            top+=1
            if top>down:
                break
            for i in range(top,down+1):
                res.append(matrix[i][right])
            right-=1
            if right<left:
                break
            for i in range(right,left-1,-1):
                res.append(matrix[down][i])
            down-=1
            for i in range(down,top-1,-1):
                res.append(matrix[i][left])
            left+=1
        return res
        
so=Solution()
matrix=[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
#print so.spiralOrder(matrix)
matrix=[
[1,2,3,4],
[5,6,7,8],
[9,10,11,12],
[13,14,15,16]
]
print((so.spiralOrder(matrix)))